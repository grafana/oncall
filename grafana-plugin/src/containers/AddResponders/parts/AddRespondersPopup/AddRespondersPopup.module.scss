.add-responders-dropdown {
  max-height: 500px;
  overflow: hidden;
  border: var(--border-medium);
  position: absolute;
  right: 16px;
  top: 55px;
  background: var(--primary-background);
  width: 340px;
  z-index: 10;
}

.info-alert {
  margin: 8px;
}

.learn-more-link {
  display: inline-block;
}

.responder-item {
  cursor: pointer;
  width: 280px;
  overflow: hidden;
}

.responders-filters {
  margin: 8px;
}

.radio-buttons {
  margin: 8px;
}

.loading-placeholder {
  margin: 8px;
}

.table {
  max-height: 150px;
  overflow: auto;
  padding: 4px 0px;

  & tr:hover {
    background: var(--background-secondary) !important;
  }

  & tbody tr:nth-child(odd) {
    background: unset;
  }
}

.user-results-section-header {
  padding: 10px 8px;
}
