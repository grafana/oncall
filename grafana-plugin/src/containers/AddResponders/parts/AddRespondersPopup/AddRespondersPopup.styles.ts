import { css } from '@emotion/css';
import { GrafanaTheme2 } from '@grafana/data';

export const getAddRespondersPopupStyles = (theme: GrafanaTheme2) => {
  return {
    addRespondersDropdown: css`
      max-height: 500px;
      overflow: hidden;
      border: 1px solid ${theme.colors.border.medium};
      position: absolute;
      right: 16px;
      top: 55px;
      background: ${theme.colors.background.primary};
      width: 340px;
      z-index: 10;
    `,

    infoAlert: css`
      margin: 8px;
    `,

    learnMoreLink: css`
      display: inline-block;
    `,

    responderItem: css`
      cursor: pointer;
      width: 280px;
      overflow: hidden;
    `,

    responderName: css`
      word-break: normal;
    `,

    responderTeam: css`
      text-align: right;
    `,

    respondersFilters: css`
      margin: 8px;
    `,

    radioButtons: css`
      margin: 8px;
    `,

    LoadingPlaceholder: css`
      margin: 8px;
    `,

    table: css`
      max-height: 150px;
      overflow: auto;
      padding: 4px 0;

      & tr:hover {
        background: ${theme.colors.background.secondary} !important;
      }

      & tbody tr:nth-child(odd) {
        background: unset;
      }
    `,

    userResultsSectionHeader: css`
      padding: 10px 8px;
    `,
  };
};
