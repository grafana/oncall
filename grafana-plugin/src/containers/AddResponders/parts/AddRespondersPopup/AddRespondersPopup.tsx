import React, { useState, useCallback, useEffect, useRef, FC } from 'react';

import { Alert, HorizontalGroup, Icon, Input, LoadingPlaceholder, RadioButtonGroup } from '@grafana/ui';
import cn from 'classnames/bind';
import { observer } from 'mobx-react';
import { ColumnsType } from 'rc-table/lib/interface';

import { Avatar } from 'components/Avatar/Avatar';
import { GTable } from 'components/GTable/GTable';
import { Text } from 'components/Text/Text';
import { GrafanaTeam } from 'models/grafana_team/grafana_team.types';
import { UserHelper } from 'models/user/user.helpers';
import { ApiSchemas } from 'network/oncall-api/api.types';
import { useStore } from 'state/useStore';
import { useDebouncedCallback, useOnClickOutside } from 'utils/hooks';

import styles from './AddRespondersPopup.module.scss';

type Props = {
  mode: 'create' | 'update';
  visible: boolean;
  setVisible: (value: boolean) => void;

  setCurrentlyConsideredUser: (user: ApiSchemas['UserIsCurrentlyOnCall']) => void;
  setShowUserConfirmationModal: (value: boolean) => void;

  existingPagedUsers?: ApiSchemas['AlertGroup']['paged_users'];
};

const cx = cn.bind(styles);

enum TabOptions {
  Teams = 'teams',
  Users = 'users',
}

export const AddRespondersPopup = observer(
  ({
    mode,
    visible,
    setVisible,
    existingPagedUsers = [],
    setCurrentlyConsideredUser,
    setShowUserConfirmationModal,
  }: Props) => {
    const { directPagingStore, grafanaTeamStore } = useStore();
    const { selectedTeamResponder, selectedUserResponders } = directPagingStore;

    const isCreateMode = mode === 'create';

    const [searchLoading, setSearchLoading] = useState(true);
    const [activeOption, setActiveOption] = useState<TabOptions>(isCreateMode ? TabOptions.Teams : TabOptions.Users);
    const [teamSearchResults, setTeamSearchResults] = useState<GrafanaTeam[]>([]);
    const [onCallUserSearchResults, setOnCallUserSearchResults] = useState<Array<ApiSchemas['UserIsCurrentlyOnCall']>>(
      []
    );
    const [notOnCallUserSearchResults, setNotOnCallUserSearchResults] = useState<
      Array<ApiSchemas['UserIsCurrentlyOnCall']>
    >([]);
    const [searchTerm, setSearchTerm] = useState('');

    const ref = useRef();

    useOnClickOutside(ref, () => {
      setVisible(false);
    });

    const handleSetSearchTerm = useCallback(
      (e: React.ChangeEvent<HTMLInputElement>) => {
        setSearchTerm(e.target.value);
      },
      [setSearchTerm]
    );

    const onClickUser = useCallback(
      async (user: ApiSchemas['UserIsCurrentlyOnCall']) => {
        if (isCreateMode && user.is_currently_oncall) {
          directPagingStore.addUserToSelectedUsers(user);
        } else {
          setCurrentlyConsideredUser(user);
          setShowUserConfirmationModal(true);
        }
        setVisible(false);
      },
      [isCreateMode, setShowUserConfirmationModal, setVisible]
    );

    const addTeamResponder = useCallback(
      (team: GrafanaTeam) => {
        setVisible(false);
        directPagingStore.updateSelectedTeam(team);

        /**
         * can't select more than one team so we mind as well auto-switch the selected tab
         * to the users section in case the user wants to come back and user(s)
         */
        setActiveOption(TabOptions.Users);
      },
      [setVisible, directPagingStore, setActiveOption]
    );

    const searchForUsers = useCallback(async () => {
      const _search = async (is_currently_oncall: boolean) => {
        const response = await UserHelper.search({
          searchTerm,
          is_currently_oncall,
        });
        return response.results;
      };

      const [onCallUserSearchResults, notOnCallUserSearchResults] = await Promise.all([_search(true), _search(false)]);

      setOnCallUserSearchResults(onCallUserSearchResults as Array<ApiSchemas['UserIsCurrentlyOnCall']>);
      setNotOnCallUserSearchResults(notOnCallUserSearchResults as Array<ApiSchemas['UserIsCurrentlyOnCall']>);
    }, [searchTerm]);

    const searchForTeams = useCallback(async () => {
      await grafanaTeamStore.updateItems(searchTerm, false, true, false);
      setTeamSearchResults(grafanaTeamStore.getSearchResult());
    }, [searchTerm]);

    const handleSearchTermChange = useDebouncedCallback(async () => {
      setSearchLoading(true);

      if (isCreateMode && activeOption === TabOptions.Teams) {
        await searchForTeams();
      } else {
        await searchForUsers();
      }

      setSearchLoading(false);
    }, 500);

    const onChangeTab = useCallback(
      async (tab: TabOptions) => {
        /**
         * there's no need to trigger a new search request when the user changes tabs if they don't have a
         * search term
         */
        if (searchTerm) {
          setSearchLoading(true);

          if (activeOption === TabOptions.Teams) {
            await searchForTeams();
          } else {
            await searchForUsers();
          }

          setSearchLoading(false);
        }

        setActiveOption(tab);
      },
      [searchTerm]
    );

    useEffect(handleSearchTermChange, [searchTerm]);

    /**
     * in the context where some user(s) have already been paged (ex. on a direct paging generated
     * alert group detail page), we should filter out the search results to not include these users
     */
    useEffect(() => {
      if (existingPagedUsers.length > 0) {
        const existingPagedUserIds = existingPagedUsers.map(({ pk }) => pk);

        const _filterUsers = (users: Array<ApiSchemas['UserIsCurrentlyOnCall']>) =>
          users.filter(({ pk }) => !existingPagedUserIds.includes(pk));

        setOnCallUserSearchResults(_filterUsers);
        setNotOnCallUserSearchResults(_filterUsers);
      }
    }, [existingPagedUsers]);

    /**
     * pre-populate the users and teams search results so that when the user opens AddRespondersPopup it is already
     * populated with data (nicer UX)
     */
    useEffect(() => {
      (async () => {
        /**
         * teams are not relevant when the component is rendered in "update" mode so we skip fetching teams here
         */
        if (isCreateMode) {
          await searchForTeams();
        }

        await searchForUsers();
        setSearchLoading(false);
      })();
    }, []);

    const userIsSelected = useCallback(
      (user: ApiSchemas['UserIsCurrentlyOnCall']) =>
        selectedUserResponders.some((userResponder) => userResponder.data.pk === user.pk),
      [selectedUserResponders]
    );

    const teamColumns: ColumnsType<GrafanaTeam> = [
      // TODO: how to make the rows span full width properly?
      {
        width: 300,
        render: (team: GrafanaTeam) => {
          const { avatar_url, name, number_of_users_currently_oncall } = team;

          return (
            <div onClick={() => addTeamResponder(team)} className={cx('responder-item')}>
              <HorizontalGroup justify="space-between">
                <HorizontalGroup>
                  <Avatar size="small" src={avatar_url} />
                  <Text>{name}</Text>
                </HorizontalGroup>
                {number_of_users_currently_oncall > 0 && (
                  <Text type="secondary">
                    {number_of_users_currently_oncall} user{number_of_users_currently_oncall > 1 ? 's' : ''} on-call
                  </Text>
                )}
              </HorizontalGroup>
            </div>
          );
        },
        key: 'Title',
      },
    ];

    const userColumns: ColumnsType<ApiSchemas['UserIsCurrentlyOnCall']> = [
      // TODO: how to make the rows span full width properly?
      {
        width: 300,
        render: (user: ApiSchemas['UserIsCurrentlyOnCall']) => {
          const { avatar, name, username, teams } = user;
          const disabled = userIsSelected(user);

          return (
            <div onClick={() => (disabled ? undefined : onClickUser(user))} className={cx('responder-item')}>
              <HorizontalGroup justify="space-between">
                <HorizontalGroup>
                  <Avatar size="small" src={avatar} />
                  <Text type={disabled ? 'disabled' : undefined}>{name || username}</Text>
                </HorizontalGroup>
                {/* TODO: we should add an elippsis and/or tooltip in the event that the user has a ton of teams */}
                {teams?.length > 0 && <Text type="secondary">{teams.map(({ name }) => name).join(', ')}</Text>}
              </HorizontalGroup>
            </div>
          );
        },
        key: 'username',
      },
      {
        width: 40,
        render: (user: ApiSchemas['UserIsCurrentlyOnCall']) => (userIsSelected(user) ? <Icon name="check" /> : null),
        key: 'Checked',
      },
    ];

    const UserResultsSection: FC<{ header: string; users: Array<ApiSchemas['UserIsCurrentlyOnCall']> }> = ({
      header,
      users,
    }) =>
      users.length > 0 && (
        <>
          <Text type="secondary" className={cx('user-results-section-header')}>
            {header}
          </Text>
          <GTable<ApiSchemas['UserIsCurrentlyOnCall']>
            emptyText={users ? 'No users found' : 'Loading...'}
            rowKey="pk"
            columns={userColumns}
            data={users}
            className={cx('table')}
            showHeader={false}
          />
        </>
      );

    return (
      visible && (
        <div data-testid="add-responders-popup" ref={ref} className={cx('add-responders-dropdown')}>
          <Input
            suffix={<Icon name="search" />}
            key="search"
            className={cx('responders-filters')}
            data-testid="add-responders-search-input"
            value={searchTerm}
            placeholder="Search"
            // @ts-ignore
            width={'unset'}
            onChange={handleSetSearchTerm}
          />
          {isCreateMode && (
            <RadioButtonGroup
              options={[
                { value: TabOptions.Teams, label: 'Teams' },
                { value: TabOptions.Users, label: 'Users' },
              ]}
              className={cx('radio-buttons')}
              value={activeOption}
              onChange={onChangeTab}
              fullWidth
            />
          )}
          {searchLoading && <LoadingPlaceholder className={cx('loading-placeholder')} text="Loading..." />}
          {!searchLoading && activeOption === TabOptions.Teams && (
            <>
              {selectedTeamResponder ? (
                <Alert
                  severity="info"
                  title="You can add only one team per escalation. Please remove the existing team before adding a new one."
                />
              ) : (
                <>
                  <Alert
                    className={cx('info-alert')}
                    severity="info"
                    title={
                      (
                        <Text type="primary">
                          You can only page teams which have a Direct Paging integration that is configured.{' '}
                          <a
                            className={cx('learn-more-link')}
                            href="https://grafana.com/docs/oncall/latest/integrations/manual/#set-up-direct-paging-for-a-team"
                            target="_blank"
                            rel="noreferrer"
                          >
                            <Text type="link">
                              <HorizontalGroup spacing="xs">
                                Learn more
                                <Icon name="external-link-alt" />
                              </HorizontalGroup>
                            </Text>
                          </a>
                        </Text>
                      ) as any
                    }
                  />
                  <GTable<GrafanaTeam>
                    emptyText={teamSearchResults ? 'No teams found' : 'Loading...'}
                    rowKey="id"
                    columns={teamColumns}
                    data={teamSearchResults}
                    className={cx('table')}
                    showHeader={false}
                  />
                </>
              )}
            </>
          )}
          {!searchLoading && activeOption === TabOptions.Users && (
            <>
              <Alert
                className={cx('info-alert')}
                severity="info"
                title={
                  (
                    <Text type="primary">
                      We display a maximum of 100 users per category. Use the search bar above to refine results. You
                      can search by username, email, or team name.
                    </Text>
                  ) as any
                }
              />
              <UserResultsSection header="On-call now" users={onCallUserSearchResults} />
              <div style={{ marginTop: '10px' }}>
                <UserResultsSection header="Not on-call" users={notOnCallUserSearchResults} />
              </div>
            </>
          )}
        </div>
      )
    );
  }
);
