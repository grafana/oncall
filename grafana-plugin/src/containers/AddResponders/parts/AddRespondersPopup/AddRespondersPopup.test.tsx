import React from 'react';

import { render } from '@testing-library/react';
import { Provider } from 'mobx-react';

import { AddRespondersPopup } from './AddRespondersPopup';

describe('AddRespondersPopup', () => {
  const teams = [
    {
      pk: 1,
      avatar_url: 'https://example.com',
      name: 'my test team',
      number_of_users_currently_oncall: 1,
    },
    {
      pk: 2,
      avatar_url: 'https://example.com',
      name: 'my test team 2',
      number_of_users_currently_oncall: 0,
    },
  ];

  test('it shows a loading message initially', () => {
    const mockStoreValue = {
      directPagingStore: {
        selectedTeamResponder: null,
      },
      grafanaTeamStore: {
        getSearchResult: jest.fn().mockReturnValue(teams),
        updateItems: jest.fn(),
      },
      userStore: {
        search: jest.fn().mockReturnValue({ results: [] }),
      },
    };

    const component = render(
      <Provider store={mockStoreValue}>
        <AddRespondersPopup
          mode="create"
          visible={true}
          setVisible={jest.fn()}
          setCurrentlyConsideredUser={jest.fn()}
          setShowUserConfirmationModal={jest.fn()}
        />
      </Provider>
    );

    expect(component.container).toMatchSnapshot();
  });
});
