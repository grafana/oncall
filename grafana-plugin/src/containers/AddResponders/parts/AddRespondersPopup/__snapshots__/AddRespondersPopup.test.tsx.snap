// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRespondersPopup it shows a loading message initially 1`] = `
<div>
  <div
    class="add-responders-dropdown"
    data-testid="add-responders-popup"
  >
    <div
      class="css-qfli5h-input-wrapper responders-filters"
      data-testid="input-wrapper"
    >
      <div
        class="css-10lnb82-input-inputWrapper"
      >
        <input
          class="css-8tk2dk-input-input"
          data-testid="add-responders-search-input"
          placeholder="Search"
          style="padding-right: 12px;"
          value=""
        />
        <div
          class="css-7099m8-input-suffix"
        >
          <div
            class="css-1j2891d-Icon"
          />
        </div>
      </div>
    </div>
    <div
      class="radio-buttons css-1nxrz2e"
      role="radiogroup"
    >
      <div
        class="css-1hvl7lx"
      >
        <input
          checked=""
          class="css-1f9hgw3"
          id="option-teams-radiogroup-1"
          name="radiogroup-1"
          type="radio"
        />
        <label
          class="css-10bka4u"
          for="option-teams-radiogroup-1"
        >
          Teams
           
        </label>
      </div>
      <div
        class="css-1hvl7lx"
      >
        <input
          class="css-1f9hgw3"
          id="option-users-radiogroup-1"
          name="radiogroup-1"
          type="radio"
        />
        <label
          class="css-10bka4u"
          for="option-users-radiogroup-1"
        >
          Users
           
        </label>
      </div>
    </div>
    <div
      class="css-1yjvs5a loading-placeholder"
    >
      Loading...
       
      <div
        class="css-1tqtz24"
        data-testid="Spinner"
      >
        <i
          aria-label="loading spinner"
          class="fa fa-spinner fa-spin fa-spin"
        />
      </div>
    </div>
  </div>
</div>
`;
